gk/main.c

```diff
@@ -2135,23 +2050,156 @@ process_pkts_back(uint16_t port_back, uint16_t port_front,
 	process_pkts_acl(&gk_conf->net->back, lcore, acl6, RTE_ETHER_TYPE_IPV6);
 }
 
+static void
+update_flow_entry(struct flow_entry *fe, struct ggu_policy *policy)
+{
+	uint64_t now = rte_rdtsc();
+
+	switch (policy->state) {
+	case GK_GRANTED:
+		fe->state = GK_GRANTED;
+		fe->u.granted.cap_expire_at = now +
+			policy->params.granted.cap_expire_sec *
+			cycles_per_sec;
+		fe->u.granted.tx_rate_kib_cycle =
+			policy->params.granted.tx_rate_kib_sec;
+		fe->u.granted.send_next_renewal_at = now +
+			policy->params.granted.next_renewal_ms *
+			cycles_per_ms;
+		fe->u.granted.renewal_step_cycle =
+			policy->params.granted.renewal_step_ms *
+			cycles_per_ms;
+		fe->u.granted.budget_renew_at = now + cycles_per_sec;
+		fe->u.granted.budget_byte =
+			(uint64_t)fe->u.granted.tx_rate_kib_cycle * 1024;
+		break;
+
+	case GK_DECLINED:
+		fe->state = GK_DECLINED;
+		fe->u.declined.expire_at = now +
+			policy->params.declined.expire_sec * cycles_per_sec;
+		break;
+
+	case GK_BPF:
+		fe->state = GK_BPF;
+		fe->u.bpf.expire_at = now +
+			policy->params.bpf.expire_sec * cycles_per_sec;
+		fe->u.bpf.program_index = policy->params.bpf.program_index;
+		fe->u.bpf.cookie = policy->params.bpf.cookie;
+		break;
+
+	default:
+		GK_LOG(ERR, "Unknown flow state %u\n", policy->state);
+		break;
+	}
+}
+
+static void
+add_ggu_policy_bulk(struct gk_add_policy **policies, int num_policies,
+	struct gk_instance *instance, struct gk_config *gk_conf)
+{
+	int i;
+	struct ip_flow *flow_arr[num_policies];
+	int done_lookups;
+	uint32_t flow_hash_val_arr[num_policies];
+	int32_t pos_arr[num_policies];
+
+	for (i = 0; i < num_policies; i++) {
+		flow_arr[i] = &policies[i]->policy.flow;
+		flow_hash_val_arr[i] = policies[i]->flow_hash_val;
+	}
+
+	done_lookups = 0;
+	while (done_lookups < num_policies) {
+		int ret;
+		int r;
+		uint32_t num_keys = num_policies - done_lookups;
+		if (num_keys > RTE_HASH_LOOKUP_BULK_MAX)
+			num_keys = RTE_HASH_LOOKUP_BULK_MAX;
+
+		ret = rte_hash_lookup_bulk_with_hash(
+			instance->ip_flow_hash_table,
+			(const void **)&flow_arr[done_lookups],
+			(hash_sig_t *)&flow_hash_val_arr[done_lookups],
+			num_keys, &pos_arr[done_lookups]);
+		if (ret != 0) {
+			GK_LOG(NOTICE,
+				"failed to find multiple keys in the hash table at lcore %u\n",
+				rte_lcore_id());
+		}
+
+		r = done_lookups + num_keys;
+		for (i = done_lookups; i < r; i++) {
```

Comment:
Moving this `for` loop after `while` loop it is within, this `for` loop runs a single time instead of multiple times accounting for `done_lookups`. It makes the code simpler to read and avoid the extra indentation.

Reply:
Ignore this comment because the following patch already addresses it.

Reply:
In fact, even in the following patch, the code would be simpler if we moved the `for` loop out of the `while` loop.